### ISSUE TITLE
numbers/digits conversion

### ISSUE THREAD

**Issue by csarron** [bheinzerling/bpemb#20]
Hi there, thanks for providing such wonderful pretrained models, I noticed the pretrained model converted all digits/numbers to 0. May I ask why? Numbers are very useful and sometime crucial for downstream NLP tasks such as question answering (e.g. when asking questions related to numbers).

**Comment by bheinzerling** [comment]
I agree that numbers are very useful for many tasks.

The problem with numbers is that there are so many of them. I trained a BPE model without replacing digits with 0. With a vocabulary size of 100k there were 9787 BPE symbols consisting of numbers. That is, almost 10 percent of the vocabulary will be taken up by numbers, which could otherwise be used for actual words.

If we replace all digits with 0, then only 1511 BPE symbols consist of numbers, a much smaller fraction of the total vocabulary size. If your task focuses on numbers then allotting 10% of the vocabulary makes sense, but for other tasks you might not want to do this.

Another problem is that even when we leave all numbers as is, there will be many numbers missing. Depending on the training corpus, we might get embeddings for, say, 1896, 1897, 1899,  and 1900, but not for 1898. Now our neural model gets input embeddings for 1896, 1897, 1899,  and 1900, but for 1898 it has to rely on composition, e.g. compose the embeddings for 18 and 98.

By replacing all digits with 0, we get only one embedding for years, i.e. the embedding for 0000, so that even if one year didn't occur in the training corpus, it will still be mapped to the same embedding.
Of course, that doesn't help if you actually want to embed the actual year, and not just "a four-digit number".

In summary, replacing digits with 0 is a trade-off in which I saw some advantages for the tasks I was interested in when training the embeddings, but yes, for other tasks it might be better to keep all numbers.

**Comment by csarron** [comment]
Got it. Thanks for the reply!

**Comment by zbyte64** [comment]
I ended up converting the numbers to their word counterparts with the following:

```
from num2words import num2words
import re

text_embed_size = 300

bpemb_en = BPEmb(lang="en", dim=text_embed_size)
word_or_num_re = r"(?P<num>[0-9]+)|(?P<word>[\w]+)"


def _chunk_to_word(chunk):
    v = chunk.groupdict()
    s = v["word"]
    if v["num"] is not None:
        s = num2words(v["num"])
    return s


def embed(x, numeral_re=r"[0-9]+"):
    if x is not None:
        if not isinstance(x, str):
            x = str(x)
        # https://github.com/bheinzerling/bpemb/issues/20
        stream = map(_chunk_to_word, re.finditer(word_or_num_re, x))
        x = " ".join(stream)
        return bpemb_en.embed(x)
```
